### PR TITLE
[FLINK-18864][python] Support key_by() operation for Python DataStrea…

### PR DESCRIPTION
**Pull request by shuiqiangchen**
## What is the purpose of the change

Support key_by() operation for Python DataStream API.

## Brief change log

- Add a new class nemed KeyStream to represent a keyed DataStream.
- Add key_by() interface for DataStream.

## Verifying this change

This change has test cases covered in test_key_by() and test_key_by_map() in test_data_stream.py

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by hequn8128** [comment]
Tests failed.

File: flink-python/pyflink/datastream/tests/test_data_stream.py [hunk 1]
```diff
@@ -140,15 +135,68 @@ def flat_map(value):
...
+        self.collect_util.collect(mapped_stream)
+        self.env.execute('key_by_test')
+        results = self.collect_util.results()
+        expected = ["<Row('a', 0)>", "<Row('b', 0)>", "<Row('c', 1)>", "<Row('d', 1)>",
+                    "<Row('e', 2)>"]
+        results.sort()
+        expected.sort()
+        self.assertEqual(expected, results)
+
+    def test_key_by_map(self):
```
Comment (hequn8128) [inline flink-python/pyflink/datastream/tests/test_data_stream.py:174]: The two tests are duplicated?

Comment (shuiqiangchen) [reply flink-python/pyflink/datastream/tests/test_data_stream.py:174]: They are almost the same, but the second test is to make sure we have not changed the original DataStream after key_by() operation with two way map operation.

Comment (hequn8128) [reply flink-python/pyflink/datastream/tests/test_data_stream.py:174]: then we can use the second one to cover the first one.

File: flink-python/pyflink/datastream/data_stream.py [hunk 2]
```diff
@@ -233,6 +234,56 @@ def flat_map(self, func: Union[Callable, FlatMapFunction], type_info: TypeInform
...
+            is_key_pickled_byte_array = True
+        generated_key_stream = KeyedStream(self.map(lambda x: (key_selector.get_key(x), x),
+                                                    type_info=Types.ROW([key_type_info,
+                                                                         output_type_info]))
+                                           ._j_data_stream
+                                           .keyBy(PickledKeySelector(is_key_pickled_byte_array)))
+        generated_key_stream._original_data_type_info = output_type_info
+        return generated_key_stream
+
+    def _align_output_type(self) -> 'DataStream':
```
Comment (hequn8128) [inline flink-python/pyflink/datastream/data_stream.py:268]: Remove this method in this PR.

File: flink-python/pyflink/datastream/data_stream.py [hunk 3]
```diff
@@ -290,3 +341,39 @@ def _get_java_python_function_operator(self, func: Union[Function, FunctionWrapp
             output_type_info.get_java_type_info(),
             j_python_data_stream_function_info)
         return j_python_data_stream_scalar_function_operator, output_type_info
+
+
+class KeyedStream(DataStream):
```
Comment (hequn8128) [inline flink-python/pyflink/datastream/data_stream.py:346]: Throw exception when perform `set_parallelism`, `name`, `uid`, etc.

File: flink-python/pyflink/datastream/data_stream.py [hunk 4]
```diff
@@ -233,6 +234,56 @@ def flat_map(self, func: Union[Callable, FlatMapFunction], type_info: TypeInform
...
+        output_type_info = typeinfo._from_java_type(j_output_type_info)
+        is_key_pickled_byte_array = False
+        if key_type_info is None:
+            key_type_info = Types.PICKLED_BYTE_ARRAY()
+            is_key_pickled_byte_array = True
+        generated_key_stream = KeyedStream(self.map(lambda x: (key_selector.get_key(x), x),
+                                                    type_info=Types.ROW([key_type_info,
+                                                                         output_type_info]))
+                                           ._j_data_stream
+                                           .keyBy(PickledKeySelector(is_key_pickled_byte_array)))
```
Comment (hequn8128) [inline flink-python/pyflink/datastream/data_stream.py:264]: .keyBy(PickledKeySelector(), key_type_info)

File: flink-python/pyflink/datastream/data_stream.py [hunk 5]
```diff
@@ -233,6 +234,54 @@ def flat_map(self, func: Union[Callable, FlatMapFunction], type_info: TypeInform
...
+            key_type_info = Types.PICKLED_BYTE_ARRAY()
+        generated_key_stream = KeyedStream(self.map(lambda x: (key_selector.get_key(x), x),
+                                                    type_info=Types.ROW([key_type_info,
+                                                                         output_type_info]))
+                                           ._j_data_stream
+                                           .keyBy(PickledKeySelector()))
+        generated_key_stream._original_data_type_info = output_type_info
+        return generated_key_stream
+
+    def _align_output_type(self) -> 'DataStream':
```
[the remark this answers is not in the document]
Comment (shuiqiangchen) [reply flink-python/pyflink/datastream/data_stream.py:268]: Ok, I will remove it in this pr.

File: flink-python/pyflink/datastream/tests/test_data_stream.py (hunk 1 above)
Comment (shuiqiangchen) [reply flink-python/pyflink/datastream/tests/test_data_stream.py]: Ok

File: flink-python/pyflink/datastream/data_stream.py (hunk 2 above)
Comment (shuiqiangchen) [reply flink-python/pyflink/datastream/data_stream.py:268]: ok

**Pull request merged**